File: app/src/main/java/com/ipd/taxiu/adapter/EvaluateAdapter.java
package com.ipd.taxiu.adapter;

import android.app.Activity;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.ipd.taxiu.R;
import com.ipd.taxiu.bean.PictureBean;
import com.ipd.taxiu.ui.activity.PhotoSelectActivity;
import com.ipd.taxiu.widget.PictureRecyclerView;
import com.ipd.taxiu.widget.RatingBar;

import java.util.List;


/**
 * Created by Miss on 2018/7/20
 */
public class EvaluateAdapter extends RecyclerView.Adapter<EvaluateAdapter.ViewHolder> {
    private RecyclerView mRecyclerView;

    private List<String> data;
    private Context mContext;

    private View VIEW_FOOTER;

    //Type
    private int TYPE_NORMAL = 1000;
    private int TYPE_FOOTER = 1002;

    private int requestCode, resultCode;
    private Intent datas;

    private int selectPosition = -1;

    public EvaluateAdapter(List<String> data, Context mContext) {
        this.data = data;
        this.mContext = mContext;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        if (viewType == TYPE_FOOTER) {
            return new ViewHolder(VIEW_FOOTER);
        } else {
            return new ViewHolder(getLayout(R.layout.item_evaluate));
        }
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, final int position) {
        if (!isFooterView(position)) {
            holder.choose_goods_pic.initTwo();
            holder.choose_goods_pic.getEvaluateAdapter().setOnClickListener(new PictureEvaluateAdapter.OnClickListener() {
                @Override
                public void onClick() {
                    selectPosition = position;
                }
            });
            if (selectPosition == position) {
                holder.choose_goods_pic.onActivityResult(requestCode, resultCode, datas);
            }
        }
    }

    @Override
    public int getItemCount() {
        int count = (data == null ? 0 : data.size());
        if (VIEW_FOOTER != null) {
            count++;
        }
        return count;
    }

    @Override
    public int getItemViewType(int position) {
        if (isFooterView(position)) {
            return TYPE_FOOTER;
        } else {
            return TYPE_NORMAL;
        }
    }

    @Override
    public void onAttachedToRecyclerView(RecyclerView recyclerView) {
        try {
            if (mRecyclerView == null && mRecyclerView != recyclerView) {
                mRecyclerView = recyclerView;
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private View getLayout(int layoutId) {
        return LayoutInflater.from(mContext).inflate(layoutId, null);
    }


    public void addFooterView(View footerView) {
        if (haveFooterView()) {
            throw new IllegalStateException("footerView has already exists!");
        } else {
            ViewGroup.LayoutParams params = new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
            footerView.setLayoutParams(params);
            VIEW_FOOTER = footerView;
            notifyItemInserted(getItemCount() - 1);

             RatingBar rating_star1,rating_star2,rating_star3;
             TextView tv_satisfaction1,tv_satisfaction2,tv_satisfaction3;

            rating_star1 = footerView.findViewById(R.id.rating_star1);
            rating_star2 = footerView.findViewById(R.id.rating_star2);
            rating_star3 = footerView.findViewById(R.id.rating_star3);

            tv_satisfaction1 = footerView.findViewById(R.id.description_satisfaction_degree);
            tv_satisfaction2 = footerView.findViewById(R.id.logistics_satisfaction_degree);
            tv_satisfaction3 = footerView.findViewById(R.id.attitude_satisfaction_degree);

            setStar(rating_star1, tv_satisfaction1);
            setStar(rating_star2, tv_satisfaction2);
            setStar(rating_star3, tv_satisfaction3);
        }
    }


    public boolean haveFooterView() {
        return VIEW_FOOTER != null;
    }

    private boolean isFooterView(int position) {
        return haveFooterView() && position == getItemCount() - 1;
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        PictureRecyclerView choose_goods_pic;
        LinearLayout linear_layout;

        public ViewHolder(View itemView) {
            super(itemView);
            choose_goods_pic = itemView.findViewById(R.id.picture_recycler_view);
            linear_layout = itemView.findViewById(R.id.linear_layout);
        }
    }

    public void setReset(int requestCode, int resultCode, Intent datas, int selectPosition) {
        this.requestCode = 0;
        this.resultCode = 0;
        this.datas = null;
        this.selectPosition = -1;

        this.requestCode = requestCode;
        this.resultCode = resultCode;
        this.datas = datas;
        this.selectPosition = selectPosition;
        notifyDataSetChanged();
    }

    public int getSelectPosition() {
        return selectPosition;
    }


    //根据星星设置评语
    private void setStar(RatingBar ratingBar, final TextView textView){
        ratingBar.setOnRatingChangeListener(new RatingBar.OnRatingChangeListener() {
            @Override
            public void onRatingChange(float ratingCount) {
                int starNum = (int) ratingCount;
                if (starNum == 1){
                    textView.setText("差评");
                }
                if (starNum == 2){
                    textView.setText("一般");
                }
                if (starNum == 3){
                    textView.setText("还行");
                }
                if (starNum == 4){
                    textView.setText("满意");
                }
                if (starNum == 5){
                    textView.setText("非常满意");
                }
            }
        });
    }
}

File: app/src/main/java/com/ipd/taxiu/adapter/EvaluateAdapter.kt
package com.ipd.taxiu.adapter

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.ipd.taxiu.R
import com.ipd.taxiu.bean.ProductBean
import com.ipd.taxiu.bean.UploadProductEvaluateBean
import com.ipd.taxiu.imageload.ImageLoader
import com.ipd.taxiu.widget.RatingBar
import kotlinx.android.synthetic.main.item_evaluate.view.*

/**
Created by Miss on 2018/8/13
 */
class EvaluateAdapter(val context: Context, private val data: List<ProductBean>?) : RecyclerView.Adapter<EvaluateAdapter.ViewHolder>() {
    private var mRecyclerView: RecyclerView? = null
    private var VIEW_FOOTER: View? = null

    //Type
    private val TYPE_NORMAL = 1000
    private val TYPE_FOOTER = 1002

    private var selectPosition = -1

    override fun onCreateViewHolder(parent: ViewGroup?, viewType: Int): ViewHolder {
        return if (viewType == TYPE_FOOTER) {
            ViewHolder(this?.VIEW_FOOTER!!)
        } else {
            ViewHolder(LayoutInflater.from(context).inflate(R.layout.item_evaluate, parent, false))
        }
    }

    override fun getItemCount(): Int {
        var count = data?.size ?: 0
        if (VIEW_FOOTER != null) {
            count++
        }
        return count
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        if (!isFooterView(position)) {
            val info = data!![position]
            holder.itemView.picture_recycler_view.init(4)
            holder.itemView.picture_recycler_view.adapter.setOnItemClickListener(object : PictureAdapter.OnItemClickListener {
                override fun choosePicture(itemPos: Int) {
                    selectPosition = position
                    holder.itemView.picture_recycler_view.adapter.choosePicture()
                }

                override fun showDeleteDialog(itemPos: Int) {
                    holder.itemView.picture_recycler_view.adapter.showDeleteDialog(itemPos)
                }

            })
            ImageLoader.loadNoPlaceHolderImg(context, info.LOGO, holder.itemView.image_view)
            holder.itemView.tv_commodity_name.text = info.PROCUCT_NAME

            holder.itemView.product_star.star = info.star
            holder.itemView.product_star.setOnRatingChangeListener {
                info.star = it
                setStar(holder.itemView.product_star, holder.itemView.tv_product_star)
            }
        }
    }

    override fun getItemViewType(position: Int): Int {
        return if (isFooterView(position)) {
            TYPE_FOOTER
        } else {
            TYPE_NORMAL
        }
    }

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView?) {
        try {
            if (mRecyclerView == null && mRecyclerView != recyclerView) {
                mRecyclerView = recyclerView
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    fun addFooterView(footerView: View) {
        if (haveFooterView()) {
            throw IllegalStateException("footerView has already exists!")
        } else {
            val params = ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)
            footerView.layoutParams = params
            VIEW_FOOTER = footerView
            notifyItemInserted(itemCount - 1)

            setStar(footerView.findViewById(R.id.rating_star1), footerView.findViewById(R.id.description_satisfaction_degree))
            setStar(footerView.findViewById(R.id.rating_star2), footerView.findViewById(R.id.logistics_satisfaction_degree))
            setStar(footerView.findViewById(R.id.rating_star3), footerView.findViewById(R.id.attitude_satisfaction_degree))
        }
    }


    fun haveFooterView(): Boolean {
        return VIEW_FOOTER != null
    }

    private fun isFooterView(position: Int): Boolean {
        return haveFooterView() && position == itemCount - 1
    }

    fun getSelectPosition(): Int = selectPosition

    //根据星星设置评语
    private fun setStar(ratingBar: RatingBar, textView: TextView) {
        ratingBar.setOnRatingChangeListener { ratingCount ->
            val starNum = ratingCount.toInt()
            if (starNum == 1) {
                textView.text = "差评"
            }
            if (starNum == 2) {
                textView.text = "一般"
            }
            if (starNum == 3) {
                textView.text = "还行"
            }
            if (starNum == 4) {
                textView.text = "满意"
            }
            if (starNum == 5) {
                textView.text = "非常满意"
            }
        }
    }

    fun getProductEvaluateInfo(callback: (errMsg: String, list: List<UploadProductEvaluateBean>?) -> Unit) {
        val list = ArrayList<UploadProductEvaluateBean>()
        data?.forEachIndexed { index, info ->
            if (mRecyclerView == null) return@forEachIndexed
            val itemView = mRecyclerView!!.layoutManager!!.findViewByPosition(index)
            val content = itemView.et_comment_content.text.toString().trim()
            val pictureList = itemView.picture_recycler_view.getPictureList()
            val star = itemView.product_star.star
            if (star == 0f) {
                callback.invoke("请对商品打分", null)
                return
            }

            var picStr = ""

            pictureList.forEach {
                if (TextUtils.isEmpty(it.url)) {
                    callback.invoke("图片暂未上传成功", null)
                    return
                }
                picStr += "${it.url};"
            }


            if (!TextUtils.isEmpty(picStr)) {
                picStr.substring(0, picStr.length - 1)
            }
            list.add(UploadProductEvaluateBean(info.ORDER_DETAIL_ID, content, picStr, 0))
        }
        callback.invoke("", list)
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)
}